### ISSUE TITLE
FocusHighlightMode does not respect mouse input

### ISSUE THREAD

**Issue by MarcusWichelmann** [flutter/flutter#82793]
The Flutter focus manager only sets the highlight mode to `FocusHighlightMode.touch` when interacting with the app using touch but not when using the mouse. When using a Flutter Web app with a mouse, this causes the focus highlighting of e.g. a button to never "go away" because the highlight mode stays at `FocusHighlightMode.traditional` forever.

The code for this was introduced with #39589, so maybe @gspencergoog could clarify, why he implemented it this way?
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/focus_manager.dart#L1648

## Steps to Reproduce

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(DemoApp());
}

class DemoApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DemoWidget(),
    );
  }
}

class DemoWidget extends StatefulWidget {
  @override
  _DemoWidgetState createState() => _DemoWidgetState();
}

class _DemoWidgetState extends State<DemoWidget> {
  FocusHighlightMode _highlightMode;

  @override
  void initState() {
    super.initState();

    _highlightMode = FocusManager.instance.highlightMode;
    FocusManager.instance.addHighlightModeListener((value) {
      setState(() {
        _highlightMode = value;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('HighlightMode Bug Demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(child: Text("Button 1"), onPressed: () {}),
            ElevatedButton(child: Text("Button 2"), onPressed: () {}),
            ElevatedButton(child: Text("Button 3"), onPressed: () {}),
            Text(_highlightMode.toString())
          ],
        ),
      ),
    );
  }
}

```

1. Run the above app with Flutter Web on a desktop browser
2. See that the first button is highlighted because it has autofocus (or select another button using the TAB-key)
3. Use the mouse to click on any other button

**Expected results:** The highlight mode should change to `touch` or maybe `mouse` and the button highlighting is no longer shown when interacting using the mouse

**Actual results:** The highlight mode stays at `traditional` and the button highlighting never goes away.

<details>
  <summary>Flutter Doctor</summary>

```
[✓] Flutter (Channel stable, 2.0.3, on Linux, locale de_DE.UTF-8)
    • Flutter version 2.0.3 at /home/marcus/flutter
    • Framework revision 4d7946a68d (vor 9 Wochen), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /home/marcus/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /home/marcus/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio
    • Android Studio at /home/marcus/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.53.2)
    • VS Code at /usr/share/code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • sdk google atv x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 89.0.4389.72

! Doctor found issues in 1 category.
```

</details>

**Comment by darshankawar** [comment]
Thanks for the detailed report and code sample. Verified and see the same behavior on latest master.


https://user-images.githubusercontent.com/67046386/118635952-ca104f80-b7f1-11eb-925a-145fe9b2d2f8.mov


<details>
<summary> flutter doctor -v </summary>

```
[✓] Flutter (Channel master, 2.3.0-2.0.pre.253, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.3.0-2.0.pre.253 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision bac5a0db3b (5 hours ago), 2021-05-18 10:19:03 +0800
    • Engine revision 3cda7f5807
    • Dart version 2.14.0 (build 2.14.0-119.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.15.4 19E2269
      darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.212

! Doctor found issues in 1 category.



```

</details>

**Comment by gspencergoog** [comment]
As far as I can tell, the video and the description of what you see happening are exactly as designed. What part of that design were you expecting to happen differently? In your view, when should the highlight mode change to `touch`?

There is no such thing as `FocusHighlightMode.mouse`. If you would prefer that your app always be in "touch" mode, you can accomplish that by changing the `FocusManager.highlightStrategy` to be `FocusHighlightStrategy.alwaysTouch`.

**Comment by MarcusWichelmann** [comment]
@gspencergoog Thank you for your response. I think maybe it's the design decision then which I question.

Let's take a look at any normal HTML webpage: When pressing the tab key, you can navigate between all the links and buttons on the page and the focused item gets a highlighting. That's what `FocusHighlightMode.traditional` does in Flutter.
![image](https://user-images.githubusercontent.com/5501487/118766397-a8a87580-b87c-11eb-8148-b50ea75cfae1.png)

Now, when you click anywhere else on the page like on another button, the focus highlighting goes away. That's the same behaviour that I would expect from a Flutter Web app:

- I use the keyboard to navigate the page --> the focused element gets highlighted
- I click anywhere on the page --> the focus highlighting goes away

That's also how it currently works on touch devices and this makes way more sense in my eyes.

Right now the Flutter Web behaviour is that I focus button 1 with the keyboard, click on button 3 and button 1 stays focused and highlighted all the time. This is not only inconsistent with how things work anywhere else but also unexpected and irritating.

> In your view, when should the highlight mode change to touch?

As soon as I start interacting with the app using the mouse again, specifically when I click anywhere. So the focus manager should change the highlight mode to `touch` not only on touch input but also on mouse input. 

Of course, `FocusHighlightMode.touch` might be an inappropriate naming four mouse input, that's why I campe up with `FocusHighlightMode.mouse`, but I'm not sure if this differentiation would have any practical benefits.

Here is a video of how it behaves on android for comparison:
https://user-images.githubusercontent.com/5501487/118769859-e6a79880-b880-11eb-8847-8afe4e817167.mp4

_In this video you can also see another bug, that the button I click on does not receive focus and therefore, when I press space to activate it, another button gets activated than that one I last clicked on. But that's probably a story for another issue... 😄_

**Comment by gspencergoog** [comment]
Ahh, ok, I think I see what you mean now. I think this is mainly a problem of terminology, and a missing feature that we're aware of, and is probably what you would like to see implemented (as would we).

The difference in terminology, I think, is that you're talking about the highlight not being shown, but the way I would say it is that you would like the focus to be removed when a non-control is selected. Does that sound right? Clicking outside of a control unfocuses the control on host platforms, it doesn't just hide the highlight.

The feature that you want is this, I think: https://github.com/flutter/flutter/issues/82515, which would move the focus somewhere else when a non-focusable widget is selected.

As for always transferring focus to a button or other control when it is clicked, while that does happen in some cases, it doesn't happen in others, and would need to be configurable for the control, with defaults based on platform and type of control. For now, it is possible to do this if you set a focus node on the control and call `focusNode.requestFocus()` in your `onPressed` handler, which is definitely harder than it should be. We should implement a more automatic version of this. Could you submit that as another issue so we can prioritize it?

**Comment by MarcusWichelmann** [comment]
Ah, I agree and understand now, why you have implemented the highlight mode this way.

A fix for #82515 together with a fix for focusing on control-interaction would mostly resolve this (will create the issue for this now). But there is one remaining special case what I'm still not sure about:

What happens when the mouse interaction is not a click on a control but instead e.g. a drag scroll of the list view that contains the buttons? Should the list view request focus to itself then as soon as it's scrolled? If not, then a highlight mode for mouse interactions would still be needed.

Current behaviour on android: https://user-images.githubusercontent.com/5501487/118792607-1ada8380-b898-11eb-8470-4ec29c654fe7.mp4
Here you can see, that the focus highlight of the grid item (`Focus` widget) disappears as soon as I scroll. I would expect the same behaviour on the web.

**Comment by yjbanov** [comment]
Let's deduplicate with https://github.com/flutter/flutter/issues/82515. If there's a distinct issue that's not covered by https://github.com/flutter/flutter/issues/82515 let's file a new one that focuses on the things not covered.
